### PR TITLE
Generalize Redis cache

### PR DESCRIPTION
**Pull request by gregorjerse**
so it can be used to cache Feature objects to speed up slow queries.

EXP-611

### CODE REVIEW

File: resolwe/flow/managers/listener/redis_cache.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                logger.exception(
+                    __(
+                        "Could not retrieve data from Redis for keys: {}",
+                        ", ".join(redis_keys),
+                    )
+                )
+                raise
+
+        to_return = list()
+        for counter, data in enumerate(cached_data):
```
Comment (dblenkus) [inline resolwe/flow/managers/listener/redis_cache.py:152]: Do you need second pass, or can you process the data in the loop above?

File: resolwe/flow/managers/listener/redis_cache.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        The query is run in a transaction.
+        """
+
+        def get_redis_data(
+            pipeline: redis.client.Pipeline,
+        ) -> dict[str, Optional[Union[str, datetime]]]:
+            """Get the redis data from the specific key."""
+
+            cached_data = dict()
+            with suppress(Exception):
```
Comment (dblenkus) [inline resolwe/flow/managers/listener/redis_cache.py:277]: Why do you handle all possible errors in the `mget` method, and just suppress them here?

File: resolwe/flow/managers/listener/redis_cache.py [hunk 3]
```diff
@@ -152,6 +165,96 @@ def mget(
...
+        """
+        redis_key = self._lock_key(model, identifiers)
+        self._redis.set(redis_key, status.value)
+        self._redis.expire(redis_key, valid_for)
+
+    def lock(
+        self,
+        model: models.Model,
+        identifiers_list: Sequence[Sequence],
+        valid_for: int = 60,
```
Comment (dblenkus) [inline resolwe/flow/managers/listener/redis_cache.py:192]: Let's increase this to 300.

File: resolwe/flow/managers/listener/plugin_interface.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    # The name of the variable that represents a plugin manager to register with.
+    plugin_manager: PluginManager
+    abstract = True
+
+    @classmethod
+    def __init_subclass__(cls: Type["Plugin"], **kwargs):
+        """Register class with the registry on initialization."""
+        super().__init_subclass__(**kwargs)
+        # The plugin is not registered when the abstract attribute is set to True.
+        should_register = cls.__dict__.get("abstract", False)
```
Comment (dblenkus) [inline resolwe/flow/managers/listener/plugin_interface.py:58]: Change this to `abstract`. Can you get the value without accessing the `__dict__` directly?

File: resolwe/flow/managers/listener/plugin_interface.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        if should_register:
```
Comment (dblenkus) [inline resolwe/flow/managers/listener/plugin_interface.py:59]: This should be `not abstract`.

File: resolwe/test/testcases/__init__.py [hunk 6]
```diff
@@ -67,6 +68,8 @@ def setUp(self):
         """Prepare environment for test."""
         super().setUp()
 
+        redis_cache.clear()
```
Comment (dblenkus) [inline resolwe/test/testcases/__init__.py:71]: Will this clear the whole Redis cache or just for the current thread?

File: resolwe/test_helpers/test_runner.py [hunk 7]
```diff
@@ -82,6 +83,7 @@ def _manager_setup():
     update_observer_constants()
 
     manager.drain_messages()
+    redis_cache.clear()
```
Comment (dblenkus) [inline resolwe/test_helpers/test_runner.py:86]: Doesn't this serve the same purpose as the clear in the previous commit?

File: docs/CHANGELOG.rst [hunk 8]
```diff
@@ -16,6 +16,7 @@ Added
 - Add ``delete_background`` method to data, entity and collection model and queryset
 - Allow ``IP`` protocol sockets to be opened in the processing container by the seccomp
   security profile
+- Add ``PluginManager`` and ``Plugin`` classes to unify plugin management
```
Comment (dblenkus) [inline docs/CHANGELOG.rst:30]: Move this under unreleased.

File: resolwe/flow/managers/listener/listener.py [hunk 9]
```diff
@@ -200,32 +88,28 @@ def get_data_fields(self, data_id: int, field_names: Union[List[str], str]) -> A
...
-        redis_cache = self._redis_cache.get_cache(data_id, fields)
-
-        # When returned fields have value None we have to read them from the
-        # database.
-        database_fields.update(
-            name for name in redis_cache if redis_cache[name] is None
-        )
+        fields = [field_names] if isinstance(field_names, str) else field_names
+        # Retrieve the cached data. If no data is cached the result is empty dict.
+        print("before cached data")
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe/flow/managers/listener/listener.py]: Removed.

File: resolwe/flow/managers/listener/listener.py [hunk 10]
```diff
@@ -200,32 +88,28 @@ def get_data_fields(self, data_id: int, field_names: Union[List[str], str]) -> A
...
+        cached_data = cache_manager.get(Data, identifiers) or {}
+        print("After cached data", cached_data)
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe/flow/managers/listener/listener.py]: Removed.

File: resolwe/flow/managers/listener/listener.py [hunk 11]
```diff
@@ -200,32 +88,28 @@ def get_data_fields(self, data_id: int, field_names: Union[List[str], str]) -> A
...
+        # Retrieve the missing fields from the database.
+        database_fields = {field for field in fields if field not in cached_data}
 
         # Make sure to only read the data if database fields are not empty.
         # Django interprets empty dict as "read all fields".
         database_data = {}
+        print("Reading from database", database_fields)
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe/flow/managers/listener/listener.py]: Removed.

File: resolwe/test_helpers/test_runner.py (hunk 7 above)
Comment (gregorjerse) [reply resolwe/test_helpers/test_runner.py]: Almost (the other one is executed more often). I have removed this one.

**Review by dblenkus — APPROVED** [review]

**Pull request merged**